beacon-chain/db/kv/state.go

```diff
@@ -125,6 +140,110 @@ func (s *Store) SaveStates(ctx context.Context, states []iface.ReadOnlyBeaconSta
 	})
 }
 
+// SaveStatesEfficient stores multiple states to the db (new schema) using the provided corresponding roots.
+func (s *Store) SaveStatesEfficient(ctx context.Context, states []iface.ReadOnlyBeaconState, blockRoots [][32]byte) error {
```

Comment:
One more thing to be mindful of is that alot of these states here are shared by other callers in other threads. This function basically rips out the validator set and then pastes it back in at the end. If any other callers try to access the validator set they might get a panic or unintentionally mutate it while its doing the saving. As a follow up PR, we can have a lock for all the states and unlock at the end of the method.

Reply:
Add a followup issue https://github.com/prysmaticlabs/prysm/issues/9262